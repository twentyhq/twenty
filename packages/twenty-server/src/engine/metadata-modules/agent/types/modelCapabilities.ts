export type ModelCapabilities = {
  webSearch?: boolean;
};
