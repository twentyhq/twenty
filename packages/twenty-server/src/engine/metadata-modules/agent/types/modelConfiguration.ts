export type ModelConfiguration = {
  webSearch?: {
    enabled: boolean;
    configuration: object;
  };
  twitterSearch?: {
    enabled: boolean;
    configuration: object;
  };
};
