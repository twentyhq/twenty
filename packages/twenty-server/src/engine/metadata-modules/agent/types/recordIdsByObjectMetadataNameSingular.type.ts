export type RecordIdsByObjectMetadataNameSingularType = Array<{
  objectMetadataNameSingular: string;
  recordIds: string[];
}>;
