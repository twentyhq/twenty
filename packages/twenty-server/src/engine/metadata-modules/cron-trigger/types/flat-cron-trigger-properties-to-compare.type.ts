import { type FlatCronTrigger } from 'src/engine/metadata-modules/cron-trigger/types/flat-cron-trigger.type';

export type FlatCronTriggerPropertiesToCompare = keyof Pick<
  FlatCronTrigger,
  'settings'
>;
