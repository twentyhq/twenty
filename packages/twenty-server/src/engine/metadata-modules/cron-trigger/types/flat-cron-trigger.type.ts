import { type WorkspaceEntity } from 'src/engine/core-modules/workspace/workspace.entity';
import { type CronTrigger } from 'src/engine/metadata-modules/cron-trigger/entities/cron-trigger.entity';
import { type FlatEntityFrom } from 'src/engine/metadata-modules/flat-entity/types/flat-entity.type';
import { type ServerlessFunctionEntity } from 'src/engine/metadata-modules/serverless-function/serverless-function.entity';
import { type ExtractRecordTypeOrmRelationProperties } from 'src/engine/workspace-manager/workspace-migration-v2/types/extract-record-typeorm-relation-properties.type';

export type CronTriggerEntityRelationProperties =
  ExtractRecordTypeOrmRelationProperties<
    CronTrigger,
    ServerlessFunctionEntity | WorkspaceEntity
  >;

export type FlatCronTrigger = FlatEntityFrom<
  CronTrigger,
  CronTriggerEntityRelationProperties
>;
