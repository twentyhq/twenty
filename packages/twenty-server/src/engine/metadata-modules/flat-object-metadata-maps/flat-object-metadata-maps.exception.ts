import {
  appendCommonExceptionCode,
  CustomException,
} from 'src/utils/custom-exception';

export class FlatObjectMetadataMapsException extends CustomException<
  keyof typeof FlatObjectMetadataMapsExceptionCode
> {}

export const FlatObjectMetadataMapsExceptionCode = appendCommonExceptionCode({
  FIELD_METADATA_NOT_FOUND: 'FIELD_METADATA_NOT_FOUND',
  OBJECT_METADATA_NOT_FOUND: 'OBJECT_METADATA_NOT_FOUND',
  FIELD_METADATA_ALREADY_EXISTS: 'FIELD_METADATA_ALREADY_EXISTS',
  OBJECT_METADATA_ALREADY_EXISTS: 'OBJECT_METADATA_ALREADY_EXISTS',
} as const);
