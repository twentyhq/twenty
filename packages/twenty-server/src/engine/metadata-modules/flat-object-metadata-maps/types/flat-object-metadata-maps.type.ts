import { type FlatObjectMetadataWithFlatFieldMaps } from 'src/engine/metadata-modules/flat-object-metadata-maps/types/flat-object-metadata-with-flat-field-metadata-maps.type';

export type FlatObjectMetadataMaps = {
  byId: Partial<Record<string, FlatObjectMetadataWithFlatFieldMaps>>;
  idByNameSingular: Partial<Record<string, string>>;
};
