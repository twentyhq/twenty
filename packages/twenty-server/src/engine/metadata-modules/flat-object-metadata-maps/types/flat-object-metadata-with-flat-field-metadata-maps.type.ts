import { type FlatFieldMetadata } from 'src/engine/metadata-modules/flat-field-metadata/types/flat-field-metadata.type';
import { type FlatObjectMetadata } from 'src/engine/metadata-modules/flat-object-metadata/types/flat-object-metadata.type';

export type FlatFieldMetadataMaps = {
  fieldsById: Partial<Record<string, FlatFieldMetadata>>;
  fieldIdByJoinColumnName: Partial<Record<string, string>>;
  fieldIdByName: Partial<Record<string, string>>;
};

export type FlatObjectMetadataWithFlatFieldMaps = FlatObjectMetadata &
  FlatFieldMetadataMaps;
