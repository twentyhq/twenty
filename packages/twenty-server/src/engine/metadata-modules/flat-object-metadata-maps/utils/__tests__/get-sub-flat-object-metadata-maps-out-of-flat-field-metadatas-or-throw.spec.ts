import { jestExpectToBeDefined } from 'test/utils/expect-to-be-defined.util.test';
import {
  eachTestingContextFilter,
  type EachTestingContext,
} from 'twenty-shared/testing';
import { FieldMetadataType } from 'twenty-shared/types';

import { getFlatFieldMetadataMock } from 'src/engine/metadata-modules/flat-field-metadata/__mocks__/get-flat-field-metadata.mock';
import { PET_FLAT_FIELDS_MOCK } from 'src/engine/metadata-modules/flat-field-metadata/__mocks__/pet-flat-fields.mock';
import { ROCKET_FLAT_FIELDS_MOCK } from 'src/engine/metadata-modules/flat-field-metadata/__mocks__/rocket-flat-fields.mock';
import { FLAT_OBJECT_METADATA_MAPS_MOCKS } from 'src/engine/metadata-modules/flat-object-metadata-maps/mocks/flat-object-metadata-maps.mock';
import { type FlatObjectMetadataMaps } from 'src/engine/metadata-modules/flat-object-metadata-maps/types/flat-object-metadata-maps.type';
import {
  type GetSubFlatObjectMetadataMapsOfSpecificFieldsOrThrowArgs,
  getSubFlatObjectMetadataMapsOutOfFlatFieldMetadatasOrThrow,
} from 'src/engine/metadata-modules/flat-object-metadata-maps/utils/get-sub-flat-object-metadata-maps-out-of-flat-field-metadatas-or-throw.util';
import { PET_FLAT_OBJECT_MOCK } from 'src/engine/metadata-modules/flat-object-metadata/__mocks__/pet-flat-object.mock';
import { ROCKET_FLAT_OBJECT_MOCK } from 'src/engine/metadata-modules/flat-object-metadata/__mocks__/rocket-flat-object.mock';
import { fromFlatObjectMetadatasToFlatObjectMetadataMaps } from 'src/engine/metadata-modules/flat-object-metadata/utils/from-flat-object-metadatas-to-flat-object-metadata-maps.util';

type GetSubObjectMetadataMapsOrThrowTestCase = {
  input: GetSubFlatObjectMetadataMapsOfSpecificFieldsOrThrowArgs;
  shouldThrow?: true;
  expected?: FlatObjectMetadataMaps;
};

describe('getSubFlatObjectMetadataMapsOutOfFlatFieldMetadatasOrThrow', () => {
  const testCases: EachTestingContext<GetSubObjectMetadataMapsOrThrowTestCase>[] =
    [
      {
        title: 'should throw when object metadata id is not found',
        context: {
          input: {
            flatFieldMetadatas: [
              getFlatFieldMetadataMock({
                objectMetadataId: 'non-existent',
                type: FieldMetadataType.TEXT,
                uniqueIdentifier: 'unique-identifier-1',
              }),
            ],
            flatObjectMetadataMaps: FLAT_OBJECT_METADATA_MAPS_MOCKS,
          },
          shouldThrow: true,
        },
      },
      {
        title: 'should extract single object metadata from maps',
        context: {
          input: {
            flatFieldMetadatas: [PET_FLAT_FIELDS_MOCK.species],
            flatObjectMetadataMaps: FLAT_OBJECT_METADATA_MAPS_MOCKS,
          },
          expected: fromFlatObjectMetadatasToFlatObjectMetadataMaps([
            {
              ...PET_FLAT_OBJECT_MOCK,
              flatFieldMetadatas: [PET_FLAT_FIELDS_MOCK.species],
            },
          ]),
        },
      },
      {
        title: 'should extract object metadata from maps with provided fields',
        context: {
          input: {
            flatFieldMetadatas: [
              PET_FLAT_FIELDS_MOCK.species,
              PET_FLAT_FIELDS_MOCK.name,
              ROCKET_FLAT_FIELDS_MOCK.attachments,
            ],
            flatObjectMetadataMaps: FLAT_OBJECT_METADATA_MAPS_MOCKS,
          },
          expected: fromFlatObjectMetadatasToFlatObjectMetadataMaps([
            {
              ...PET_FLAT_OBJECT_MOCK,
              flatFieldMetadatas: [
                PET_FLAT_FIELDS_MOCK.name,
                PET_FLAT_FIELDS_MOCK.species,
              ],
            },
            {
              ...ROCKET_FLAT_OBJECT_MOCK,
              flatFieldMetadatas: [ROCKET_FLAT_FIELDS_MOCK.attachments],
            },
          ]),
        },
      },
    ];

  const filteredTestCases = eachTestingContextFilter(testCases);

  it.each(filteredTestCases)(
    '$title',
    ({
      context: {
        input: { flatFieldMetadatas, flatObjectMetadataMaps },
        shouldThrow = false,
        expected,
      },
    }) => {
      if (shouldThrow) {
        expect(() =>
          getSubFlatObjectMetadataMapsOutOfFlatFieldMetadatasOrThrow({
            flatFieldMetadatas,
            flatObjectMetadataMaps,
          }),
        ).toThrowErrorMatchingSnapshot();
      } else {
        jestExpectToBeDefined(expected);
        const result =
          getSubFlatObjectMetadataMapsOutOfFlatFieldMetadatasOrThrow({
            flatFieldMetadatas,
            flatObjectMetadataMaps,
          });

        expect(result).toEqual(expected);
      }
    },
  );
});
