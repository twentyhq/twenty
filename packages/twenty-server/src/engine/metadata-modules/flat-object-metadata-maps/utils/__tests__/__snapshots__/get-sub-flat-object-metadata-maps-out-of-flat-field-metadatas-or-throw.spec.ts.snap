// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSubFlatObjectMetadataMapsOutOfFlatFieldMetadatasOrThrow should throw when object metadata id is not found 1`] = `"getSubFlatObjectMetadataMapsOfSpecificFieldsOrThrow object metadata not found"`;
